<center> <font size =6> Predict Parkinson's Disease </font> </center>

Load the data.

In [1]:
import pandas as pd
df=pd.read_csv("Parkinson_dataset.csv")
df

name  MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
0    phon_R01_S01_1      119.992       157.302        74.997         0.00784   
1    phon_R01_S01_2      122.400       148.650       113.819         0.00968   
2    phon_R01_S01_3      116.682       131.111       111.555         0.01050   
3    phon_R01_S01_4      116.676       137.871       111.366         0.00997   
4    phon_R01_S01_5      116.014       141.781       110.655         0.01284   
..              ...          ...           ...           ...             ...   
190  phon_R01_S50_2      174.188       230.978        94.261         0.00459   
191  phon_R01_S50_3      209.516       253.017        89.488         0.00564   
192  phon_R01_S50_4      174.688       240.005        74.287         0.01360   
193  phon_R01_S50_5      198.764       396.961        74.904         0.00740   
194  phon_R01_S50_6      214.289       260.277        77.973         0.00567   

     MDVP:Jitter(Abs)  MDVP:RAP  MDVP:PPQ  Jitter:DDP  MDVP:Shimmer  ...  \
0             0.00007   0.00370   0.00554     0.01109       0.04374  ...   
1             0.00008   0.00465   0.00696     0.01394       0.06134  ...   
2             0.00009   0.00544   0.00781     0.01633       0.05233  ...   
3             0.00009   0.00502   0.00698     0.01505       0.05492  ...   
4             0.00011   0.00655   0.00908     0.01966       0.06425  ...   
..                ...       ...       ...         ...           ...  ...   
190           0.00003   0.00263   0.00259     0.00790       0.04087  ...   
191           0.00003   0.00331   0.00292     0.00994       0.02751  ...   
192           0.00008   0.00624   0.00564     0.01873       0.02308  ...   
193           0.00004   0.00370   0.00390     0.01109       0.02296  ...   
194           0.00003   0.00295   0.00317     0.00885       0.01884  ...   

     Shimmer:DDA      NHR     HNR  status      RPDE       DFA   spread1  \
0        0.06545  0.02211  21.033       1  0.414783  0.815285 -4.813031   
1        0.09403  0.01929  19.085       1  0.458359  0.819521 -4.075192   
2        0.08270  0.01309  20.651       1  0.429895  0.825288 -4.443179   
3        0.08771  0.01353  20.644       1  0.434969  0.819235 -4.117501   
4        0.10470  0.01767  19.649       1  0.417356  0.823484 -3.747787   
..           ...      ...     ...     ...       ...       ...       ...   
190      0.07008  0.02764  19.517       0  0.448439  0.657899 -6.538586   
191      0.04812  0.01810  19.147       0  0.431674  0.683244 -6.195325   
192      0.03804  0.10715  17.883       0  0.407567  0.655683 -6.787197   
193      0.03794  0.07223  19.020       0  0.451221  0.643956 -6.744577   
194      0.03078  0.04398  21.209       0  0.462803  0.664357 -5.724056   

      spread2        D2       PPE  
0    0.266482  2.301442  0.284654  
1    0.335590  2.486855  0.368674  
2    0.311173  2.342259  0.332634  
3    0.334147  2.405554  0.368975  
4    0.234513  2.332180  0.410335  
..        ...       ...       ...  
190  0.121952  2.657476  0.133050  
191  0.129303  2.784312  0.168895  
192  0.158453  2.679772  0.131728  
193  0.207454  2.138608  0.123306  
194  0.190667  2.555477  0.148569  

[195 rows x 24 columns]

The task associated with this dataset is to predict the presence of Parkinson's disease. Let's see how many positive cases, i.e., patients with Parkinson's, there are in the datast.

In [2]:
len(df[df["status"]==1])

147

There are 147 positive cases. The entire dataset has 195 instances. So, there are 48 negative cases (patients with no Parkinson's).

The entire dataset has 24 columns. Excluding the target column "status", there are 23 potential predictor variables. The "name" column has no predictive value. So, we are going to drop that column. There are still 22 potential predictor variables. Some of them may not be relevant. So, we are going to run a feature selection algorithm and compute the feature importance scores for all 22 predictor variables. 

In [3]:
df=df.drop(["name"],axis=1)
X=df.drop(["status"],axis=1)
X.head()

MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  MDVP:Jitter(Abs)  \
0      119.992       157.302        74.997         0.00784           0.00007   
1      122.400       148.650       113.819         0.00968           0.00008   
2      116.682       131.111       111.555         0.01050           0.00009   
3      116.676       137.871       111.366         0.00997           0.00009   
4      116.014       141.781       110.655         0.01284           0.00011   

   MDVP:RAP  MDVP:PPQ  Jitter:DDP  MDVP:Shimmer  MDVP:Shimmer(dB)  ...  \
0   0.00370   0.00554     0.01109       0.04374             0.426  ...   
1   0.00465   0.00696     0.01394       0.06134             0.626  ...   
2   0.00544   0.00781     0.01633       0.05233             0.482  ...   
3   0.00502   0.00698     0.01505       0.05492             0.517  ...   
4   0.00655   0.00908     0.01966       0.06425             0.584  ...   

   MDVP:APQ  Shimmer:DDA      NHR     HNR      RPDE       DFA   spread1  \
0   0.02971      0.06545  0.02211  21.033  0.414783  0.815285 -4.813031   
1   0.04368      0.09403  0.01929  19.085  0.458359  0.819521 -4.075192   
2   0.03590      0.08270  0.01309  20.651  0.429895  0.825288 -4.443179   
3   0.03772      0.08771  0.01353  20.644  0.434969  0.819235 -4.117501   
4   0.04465      0.10470  0.01767  19.649  0.417356  0.823484 -3.747787   

    spread2        D2       PPE  
0  0.266482  2.301442  0.284654  
1  0.335590  2.486855  0.368674  
2  0.311173  2.342259  0.332634  
3  0.334147  2.405554  0.368975  
4  0.234513  2.332180  0.410335  

[5 rows x 22 columns]

In [4]:
y=df[["status"]]
y.head()

status
0       1
1       1
2       1
3       1
4       1

Compute the feature importance scores for each of the 22 predictor variables.

In [5]:
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import mutual_info_classif

# feature extraction. We are going to output the selection scores for all features and select the features with the highest scores.
test = SelectKBest(score_func=mutual_info_classif, k="all")
fit = test.fit(X, y)
features = fit.transform(X)
# summarize scores
set_printoptions(precision=3)

scores=fit.scores_
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores.sort_values(by=["scores"],ascending=True)

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/utils/validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


scores
20  0.017913
16  0.037003
7   0.093524
17  0.095116
5   0.098027
8   0.099446
6   0.105351
13  0.110436
10  0.110558
11  0.116305
3   0.118365
9   0.121471
14  0.122338
15  0.132462
1   0.140713
4   0.160015
2   0.174088
12  0.178956
19  0.192002
0   0.206593
18  0.221368
21  0.253982

All features are relevant. We are going to use 10 of them. So, let's modify the code block above and output only the 10 features with the highest feature importance scores.

In [6]:
scores.sort_values(by=["scores"],ascending=True).nlargest(10,"scores")

scores
21  0.253982
18  0.221368
0   0.206593
19  0.192002
12  0.178956
2   0.174088
4   0.160015
1   0.140713
15  0.132462
14  0.122338

So, our predictor set is:

In [7]:
X=X.iloc[:,[14,15,1,2,4,12,19,0,18,21]]
X

NHR     HNR  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(Abs)  MDVP:APQ  \
0    0.02211  21.033       157.302        74.997           0.00007   0.02971   
1    0.01929  19.085       148.650       113.819           0.00008   0.04368   
2    0.01309  20.651       131.111       111.555           0.00009   0.03590   
3    0.01353  20.644       137.871       111.366           0.00009   0.03772   
4    0.01767  19.649       141.781       110.655           0.00011   0.04465   
..       ...     ...           ...           ...               ...       ...   
190  0.02764  19.517       230.978        94.261           0.00003   0.02745   
191  0.01810  19.147       253.017        89.488           0.00003   0.01879   
192  0.10715  17.883       240.005        74.287           0.00008   0.01667   
193  0.07223  19.020       396.961        74.904           0.00004   0.01588   
194  0.04398  21.209       260.277        77.973           0.00003   0.01373   

      spread2  MDVP:Fo(Hz)   spread1       PPE  
0    0.266482      119.992 -4.813031  0.284654  
1    0.335590      122.400 -4.075192  0.368674  
2    0.311173      116.682 -4.443179  0.332634  
3    0.334147      116.676 -4.117501  0.368975  
4    0.234513      116.014 -3.747787  0.410335  
..        ...          ...       ...       ...  
190  0.121952      174.188 -6.538586  0.133050  
191  0.129303      209.516 -6.195325  0.168895  
192  0.158453      174.688 -6.787197  0.131728  
193  0.207454      198.764 -6.744577  0.123306  
194  0.190667      214.289 -5.724056  0.148569  

[195 rows x 10 columns]

For this dataset, I am not going to implement any outlier removal algorithm. This is a small dataset. I need all the data to do a meaningful analysis and train the model.

Let's standardize the predictor variables.

In [8]:
from sklearn import preprocessing
names=X.columns
scaler=preprocessing.StandardScaler()
X=scaler.fit_transform(X)
X=pd.DataFrame(X, columns=names)
X.head()

NHR       HNR  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(Abs)  MDVP:APQ  \
0 -0.067893 -0.193225     -0.436165     -0.952037          0.749759  0.332985   
1 -0.137843 -0.634508     -0.530974     -0.057721          1.037674  1.159454   
2 -0.291633 -0.279760     -0.723168     -0.109875          1.325589  0.699187   
3 -0.280719 -0.281346     -0.649092     -0.114229          1.325589  0.806859   
4 -0.178026 -0.506745     -0.606245     -0.130608          1.901418  1.216839   

    spread2  MDVP:Fo(Hz)   spread1       PPE  
0  0.480477    -0.829300  0.801323  0.868886  
1  1.311185    -0.770972  1.479853  1.803605  
2  1.017682    -0.909476  1.141445  1.402661  
3  1.293840    -0.909622  1.440945  1.806954  
4  0.096195    -0.925657  1.780940  2.267082

One-hot encode the target variable.

In [9]:
from sklearn.preprocessing import OneHotEncoder
enc=OneHotEncoder(handle_unknown="ignore")
y=pd.DataFrame(enc.fit_transform(y).toarray())
y.head()

0    1
0  0.0  1.0
1  0.0  1.0
2  0.0  1.0
3  0.0  1.0
4  0.0  1.0

Split the predictor variables and the target variable into train test subsets.

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.15,random_state=42)

Build the neural network model, train it and test it.

In [11]:
length=X_train.shape[1]
num_classes=y_test.shape[1]

# Load the libraries
import keras
from keras.models import Sequential
from keras.layers import Dense

# define classification model

def classification_model():
    # create model
    model = Sequential()
    model.add(Dense(length, activation='relu', input_shape=(length,)))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation= 'relu'))
    model.add(Dense(50, activation = 'relu'))
    model.add(Dense(num_classes, activation='softmax'))
    
    
    # compile model
    model.compile(optimizer='Adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

In [19]:
model=classification_model()
model.fit(X_train,y_train,validation_data=(X_test, y_test),epochs=18, verbose=2)
model.evaluate(X_test,y_test,verbose=0)

Train on 165 samples, validate on 30 samples
Epoch 1/18
 - 2s - loss: 0.6633 - acc: 0.7758 - val_loss: 0.6065 - val_acc: 0.8667
Epoch 2/18
 - 0s - loss: 0.5958 - acc: 0.7333 - val_loss: 0.5210 - val_acc: 0.8667
Epoch 3/18
 - 0s - loss: 0.5291 - acc: 0.7333 - val_loss: 0.4262 - val_acc: 0.8667
Epoch 4/18
 - 0s - loss: 0.4599 - acc: 0.7333 - val_loss: 0.3462 - val_acc: 0.9000
Epoch 5/18
 - 0s - loss: 0.4107 - acc: 0.7939 - val_loss: 0.2963 - val_acc: 0.9333
Epoch 6/18
 - 0s - loss: 0.3858 - acc: 0.8424 - val_loss: 0.2770 - val_acc: 0.9333
Epoch 7/18
 - 0s - loss: 0.3637 - acc: 0.8667 - val_loss: 0.2685 - val_acc: 0.9333
Epoch 8/18
 - 0s - loss: 0.3392 - acc: 0.8667 - val_loss: 0.2615 - val_acc: 0.9333
Epoch 9/18
 - 0s - loss: 0.3194 - acc: 0.8788 - val_loss: 0.2471 - val_acc: 0.9333
Epoch 10/18
 - 0s - loss: 0.3049 - acc: 0.8788 - val_loss: 0.2409 - val_acc: 0.9333
Epoch 11/18
 - 0s - loss: 0.2925 - acc: 0.8788 - val_loss: 0.2413 - val_acc: 0.9000
Epoch 12/18
 - 0s - loss: 0.2770 - acc: 

[0.23810307681560516, 0.9333333373069763]

The neural network model has a 93.33% accuracy on test data. That is pretty good.

Compare the actual labels and predicted labels and compute the confusion a matrix.

The neural network model above is a multilabel classification model. The confusion_matrix function from "from sklearn.metrics import confusion_matrix" could not handle a multilable output. And I got an import error every time when I ran "from sklearn.metrics import multilable_confusion_matrix", for which I had not found a workable solution. So, I am going to mannually compute the confusion matrix. That would be a good exercise.

Join the actual labels and predicted labels.

In [72]:
import numpy as np
y_pred=pd.DataFrame(model.predict(X_test)).rename(columns={0:"pred_0",1:"pred_1"})

y_test=pd.DataFrame(np.asarray(y_test)).rename(columns={0:"actual_0",1:"actual_1"})

label=pd.concat([y_test,y_pred],axis=1)

label.head()

actual_0  actual_1    pred_0    pred_1
0       0.0       1.0  0.000002  0.999998
1       0.0       1.0  0.003277  0.996723
2       0.0       1.0  0.001163  0.998837
3       0.0       1.0  0.090905  0.909095
4       0.0       1.0  0.003971  0.996029

We see that the predicted outputs of the model are nothing like the actual binary labels. That is because the model computed the probability that an instance would belong to a particular class. For first instance, the probability of it belonging to Class "0" (no Parkinson's) is 0.000002, or 0.0002%, while the probability of it belonging to Class "1" (Parkinson's) is 0.999998, or 99.9998%. So, the first instance belongs to Class 1. Using a threshold of 0.5, probabilities below 50% will be classified as non-presence of that class. 

In [88]:
import numpy as np
pred_1=np.where(label["pred_1"]>=0.5)
print("The row indices of the predicted positive cases: ",pred_1)
print("The number of positive cases predicted by the model: ", len(np.asarray(pred_1).T))

The row indices of the predicted positive cases:  (array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 24, 25, 26, 27, 29]),)
The number of positive cases predicted by the model:  28


The model predicted 28 positive cases. Let's see how many actual cases there are in the test data.

In [97]:
print("There are ",len(label[label["actual_1"]==1])," actual positive cases in the test data.")

There are  26  actual positive cases in the test data.


Let's pick out the actual positive cases and see how that compare to the predicted positive cases. We want to see the model's sensitivity. How well did the model performed on the predicting the cases for which the patients did have Parksinson's disease?

In [91]:
sub_label_1=label[label["actual_1"]==1]
missed_sub_label_1=sub_label_1[sub_label_1["pred_1"]<0.5]
missed_sub_label_1

Empty DataFrame
Columns: [actual_0, actual_1, pred_0, pred_1]
Index: []

The first line of the above code block picked out the actual positive cases, formed a new dataframe with only instances. Then, the second line examined whether there  is any case in this new dataframe that the model had predicted to be negative. The code block output nothing. That means the model correctly predicted ALL positive cases. It correclty predicted all the instances that the patients had Parkinson's disease. There are a total of 26 positive cases in the test data. The model predicted 28. So, True Positive is 26 and False Positive is 2.

Let's also see how the model performed on the negative cases in the test data.

In [80]:
pred_0=np.where(label["pred_0"]>=0.5)
pred_0

(array([23, 28]),)

The model only predicted 2 negative cases. Let's see how that compared to the actual number of negative cases in the test data.

In [81]:
sub_label_0=label[label["actual_0"]==1]
sub_label_0

actual_0  actual_1    pred_0    pred_1
10       1.0       0.0  0.008427  0.991573
18       1.0       0.0  0.414245  0.585755
23       1.0       0.0  0.961068  0.038932
28       1.0       0.0  0.886173  0.113827

There are 4 negative cases in the test data. The model predicted 2. Hence, True Negative is 2. 

We got our confusion matrix.

In [98]:
print("Confusion matrix: \n", np.array([[26,0],[2,2]]))

Confusion matrix: 
 [[26  0]
 [ 2  2]]


On test data, the model has an accuracy of 93.33%, sensitivity of 100%, specificity of 50%. Specificity appears to be low. It is actually not a bad result, considering that there are only 4 actual negative cases in the test data and the model correctly predicted 2 of them. For patients who developed Parkinson's disease, being able to take early action will improve the quality of life for the patients as well as easing the burden of caregiving for the family members. For this purpose, this is indeed a very good model: It correctly predicted all the cases that the patients developed Parkinson's disease.

So, that is a neural network predictive model for the dataset that was featured on UCI's website this morning. 